theano/gpuarray/opt.py

```diff
@@ -895,7 +903,34 @@ def local_gpua_subtensor(op, context_name, inputs, outputs):
                                               for v in n.inputs + n.outputs])
                         for n, _ in clients[0]]):
                     return
+    # Here is the condition for the GraphToGPU opt. inputs is the
+    # inputs we want to use for the new node
+    if (x.owner and isinstance(x.owner.op, GpuFromHost)):
+        cpu_x = x.owner.inputs[0]
+        # And it is a shared var or an input of the graph.
+        # and is used by only 1 node.
+        # x is in the new graph, so we can't tests its number of clients.
+        if not cpu_x.owner and len(cpu_x.clients) == 1:
+            return
+    return GpuSubtensor(op.idx_list)
+
 
+@register_opt2([tensor.Subtensor], 'fast_compile')
+def local_gpua_subtensor_graph(op, context_name, inputs, outputs):
+    # We need different code as the condition is different as inputs
+    # aren't the same.
```

Comment:
This inconsistency among the Local optimizers recieving different inputs was my concern yesterday. Doesn't all the other local opts also receive different inputs ? (Because the inputs that we pass is 
`[mappings[i] for i in node.inputs]` )
 That is why i wanted to have the check on all the inputs of the graph inside these local opts. If the input is on the GPU, then the transfer of the entire node will be taken care of by the new opt.